File: SchoolWebApp-API/Project.Core/Entities/Staff/StaffDetails.cs
﻿using SchoolWebApp.Core.Entities.Academics;
using SchoolWebApp.Core.Entities.School;
using SchoolWebApp.Core.Entities.Settings;
using SchoolWebApp.Core.Entities.Shared;
using System.ComponentModel.DataAnnotations;

namespace SchoolWebApp.Core.Entities.Staff
{
    public class StaffDetails : Person
    {
        [Display(Name = "School identity number")]
        [StringLength(255)]
        public string? IdNumber { get; set; }

        public string? NhifNo { get; set; }
        public string? NssfNo { get; set; }
        public string? KraPinNo { get; set; }
        public DateTime? EmploymentDate { get; set; }
        public DateTime? EndofEmploymentDate { get; set; }
        public bool CurrentlyEmployed { get; set; }

        public int StaffCategoryId { get; set; }
        public StaffCategory? StaffCategory { get; set; }
        public int DesignationId { get; set; }
        public Designation? Designation { get; set; }
        public int EmploymentTypeId { get; set; }
        public EmploymentType? EmploymentType { get; set; }

        public List<StaffSubject> StaffSubjects { get; set; } = new();
        public List<StaffAttendance> StaffAttendances { get; set; } = new();
        public List<StaffDiscipline> StaffDisciplines { get; set; } = new();
        public List<Department> Departments { get; set; } = new();
        public List<Subject> Subjects { get; set; } = new();
        public List<ToDoList> ToDoLists { get; set; } = new();
    }
}

File: SchoolWebApp-Client/src/app/staff/models/staff-details.ts
import {Status} from '@/core/enums/status';
import {ResourceModel} from '@/core/models/ResourceModel';
import { Designation } from '@/settings/models/designation';
import { EmploymentType } from '@/settings/models/employment-type';
import {Gender} from '@/settings/models/gender';
import {Nationality} from '@/settings/models/nationality';
import {Religion} from '@/settings/models/religion';
import { StaffCategory } from '@/settings/models/staff-category';

export class StaffDetails extends ResourceModel<StaffDetails> {
    //Core details
    public staffImageAsBase64?: string;    
    public fullName?: string;    
    public status?: Status;
    public nationalityId?: number;
    public nationality?: Nationality;
    public religionId?: number;
    public religion?: Religion;
    public genderId?: number;
    public gender?: Gender;
    

    //Personal details
    public dateOfBirth?: Date;
    public upi?: string;
    public idNumber?: string;
    public nhifNo?: string;
    public nssfNo?: string;
    public kraPinNo?: string;

    //Employment details
    public employmentDate?: Date;
    public endofEmploymentDate?: Date;
    public currentlyEmployed?: boolean;
    public staffCategoryId?: number;
    public staffCategory?: StaffCategory;
    public designationId?: number;
    public designation?: Designation;
    public employmentTypeId?: number;
    public employmentType?: EmploymentType;

    //Contact details
    public address?: string;
    public phoneNumber?: string;
    public email?: string;
    public otherDetails?: string;

    public isSelected?: boolean = false;
    public isOriginallySelected?: boolean = false;

    public remarks?: string = '';
    public hasRecord?: boolean = false;

    public timeIn? = { hour: 8, minute: 0 };
    public timeOut? = { hour: 17, minute: 0 };

    constructor(model?: Partial<StaffDetails>) {
        super(model);
    }
}
